apps/reaper/lib/reaper/xml/xml_stream/sax_handler.ex

```diff
@@ -0,0 +1,133 @@
+defmodule XMLStream.SaxHandler do
+  @moduledoc false
+
+  defmodule State do
+    @moduledoc false
+    @enforce_keys [:tag_path, :emitter]
+    defstruct stack: [], accumulate: false, tag_stack: [], tag_path: [], emitter: nil
+
+    def new(fields \\ %{}) do
+      struct!(__MODULE__, fields)
+    end
+
+    def update(%__MODULE__{}=state, fields) do
+      struct!(state, fields)
+    end
+  end
+
+  @behaviour Saxy.Handler
+  @file_chunk_size 40_000
+
+  ######################
+  ## Client Functions ##
+  ######################
+  def start_stream(filepath, tag_path_string, callback) when is_function(callback, 1) do
+    tag_path = parse_selector(tag_path_string)
+
+    filepath
+    |> File.stream!([], @file_chunk_size)
+    |> handle_ufeff()
+    |> Saxy.parse_stream(
+      __MODULE__,
+      __MODULE__.State.new(tag_path: tag_path, emitter: callback)
+    )
+  end
+
+  ###############
+  ## Callbacks ##
+  ###############
+  def handle_event(:start_document, _, %State{}=state) do
+    ok(state)
+  end
+
+  def handle_event(:end_document, _, %State{}=state) do
+    ok(state)
+  end
+
+  def handle_event(:start_element, {tag_name, attributes}, %State{}=state) do
+    state = State.update(state, tag_stack: [tag_name | state.tag_stack])
+
+    if state.tag_stack == state.tag_path or state.accumulate do
+      tag = {tag_name, attributes, []}
+
+      state
+      |> State.update(stack: [tag | state.stack], accumulate: true)
+      |> ok()
+    else
+      ok(state)
+    end
+  end
+
+  def handle_event(:characters, chars, %State{stack: stack, accumulate: accumulate} = state) do
+    if accumulate do
```

Comment:
Can you pattern match the falsey case out in a function head and get rid of this `if/else`?

Reply:
Good catch. That makes sense for this one.